controllers/core/openstackversion_controller.go

```diff
@@ -0,0 +1,310 @@
+		       instance.Status.Conditions.MarkTrue(
+		           condition.ReadyCondition, condition.ReadyMessage)
+		   } else {
+		       // something is not ready so reset the Ready condition
+		       instance.Status.Conditions.MarkUnknown(
+		           condition.ReadyCondition, condition.InitReason, condition.ReadyInitMessage)
+		       // and recalculate it based on the state of the rest of the conditions
+		       instance.Status.Conditions.Set(
+		           instance.Status.Conditions.Mirror(condition.ReadyCondition))
+		   }*/
+
+		err := helper.PatchInstance(ctx, instance)
+		if err != nil {
+			_err = err
+			return
+		}
+	}()
+	instance.Status.AvailableVersion = envOpenStackVersion
+	instance.Status.AvailableServices = make([]string, 0)
+	for service, _ := range envVarsFromMain {
+		instance.Status.AvailableServices = append(instance.Status.AvailableServices, (strings.TrimPrefix(strings.TrimSuffix(service, "_IMAGE_URL_DEFAULT"), "RELATED_IMAGE_")))
+	}
+
+	if len(instance.Spec.TargetVersion) > 0 && instance.Spec.TargetVersion == instance.Status.TargetVersion {
+		return ctrl.Result{}, nil
+	}
+
+	// lookup the named OpenStackControlplane
+	controlPlane, err := r.lookupOpenStackControlPlane(ctx, instance)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// services
+	glance := CheckGlanceImages(ctx, Log, &controlPlane.Spec.Glance.Template, instance.Spec.ServiceExcludes, envVarsFromMain)
+	if !glance {
+		instance.Status.ServicesNeedingUpdates = appendServiceIfMissing(instance.Status.ServicesNeedingUpdates, corev1beta1.OpenStackService{ServiceName: "glance"})
+	}
+
+	cinder := CheckCinderImages(ctx, Log, &controlPlane.Spec.Cinder.Template, instance.Spec.ServiceExcludes, instance.Spec.CinderVolumeExcludes, envVarsFromMain)
+	if !cinder {
+		instance.Status.ServicesNeedingUpdates = appendServiceIfMissing(instance.Status.ServicesNeedingUpdates, corev1beta1.OpenStackService{ServiceName: "cinder"})
+	}
+
+	if instance.Spec.TargetVersion != envOpenStackVersion {
+		return ctrl.Result{}, fmt.Errorf(fmt.Sprintf("Target version must equal: %s", envOpenStackVersion))
+	}
+
+	if len(instance.Status.ServicesNeedingUpdates) > 0 {
+
+		if !glance {
+			UpdateGlanceImages(ctx, Log, &controlPlane.Spec.Glance.Template, instance.Spec.ServiceExcludes, envVarsFromMain)
+			glanceService := corev1beta1.OpenStackService{ServiceName: "glance"}
+			instance.Status.DeployedVersions = appendServiceIfMissing(instance.Status.DeployedVersions, glanceService)
+			instance.Status.ServicesNeedingUpdates = removeService(instance.Status.ServicesNeedingUpdates, glanceService)
+		}
+		if !cinder {
+			UpdateCinderImages(ctx, Log, &controlPlane.Spec.Cinder.Template, instance.Spec.ServiceExcludes, instance.Spec.CinderVolumeExcludes, envVarsFromMain)
+			cinderService := corev1beta1.OpenStackService{ServiceName: "cinder"}
+			instance.Status.DeployedVersions = appendServiceIfMissing(instance.Status.DeployedVersions, cinderService)
```

Comment:
If I understand correctly this tracks the fact that cinder needs update / updated but the serviceExclude might excluded just CindeScheduler but not excluded CinderAPI. So there is a granularity mismatch